web/packages/teleport/src/Bots/Details/BotDetails.test.tsx

```diff
@@ -0,0 +1,247 @@
+
+import darkTheme from 'design/theme/themes/darkTheme';
+import { ConfiguredThemeProvider } from 'design/ThemeProvider';
+import {
+  fireEvent,
+  render,
+  screen,
+  testQueryClient,
+  waitForElementToBeRemoved,
+  within,
+} from 'design/utils/testing';
+import { InfoGuidePanelProvider } from 'shared/components/SlidingSidePanel/InfoGuide';
+
+import { ContextProvider } from 'teleport/index';
+import { createTeleportContext } from 'teleport/mocks/contexts';
+import { defaultAccess, makeAcl } from 'teleport/services/user/makeAcl';
+import { getBotError, getBotSuccess } from 'teleport/test/helpers/bots';
+
+import { BotDetails } from './BotDetails';
+
+jest.mock('react-router', () => {
+  const actual = jest.requireActual('react-router');
+  return {
+    ...actual,
+    useHistory: jest.fn(),
+    useParams: jest.fn(() => ({
+      name: 'test-bot-name',
+    })),
+  };
+});
+
+const server = setupServer();
+
+beforeEach(() => {
+  server.listen();
+});
+
+afterEach(async () => {
+  server.resetHandlers();
+  await testQueryClient.resetQueries();
+
+  jest.clearAllMocks();
+});
+
+afterAll(() => server.close());
+
+describe('BotDetails', () => {
+  it('should show a page error state', async () => {
+    withFetchError();
+    renderComponent();
+    await waitForLoading();
+
+    expect(screen.getByText('Error: something went wrong')).toBeInTheDocument();
+  });
+
+  it('should show a not found error state', async () => {
+    withFetchError(404, 'not_found');
+    renderComponent();
+    await waitForLoading();
+
+    expect(
+      screen.getByText('Bot test-bot-name does not exist')
+    ).toBeInTheDocument();
+  });
+
+  it('should allow back navigation', async () => {
+    const goBack = jest.fn();
+    jest.mocked(useHistory).mockImplementation(
+      () =>
+        ({
+          goBack,
+        }) as unknown as ReturnType<typeof useHistory>
+    );
+
+    withFetchSuccess();
+    renderComponent();
+    await waitForLoading();
+
+    const backButton = screen.getByLabelText('back');
+    fireEvent.click(backButton);
+
+    expect(goBack).toHaveBeenCalledTimes(1);
+  });
+
+  it('should show page title', async () => {
+    withFetchSuccess();
+    renderComponent();
+    await waitForLoading();
+
+    const pageHeader = screen.getByTestId('page-header');
+    expect(pageHeader).toBeInTheDocument();
+
+    expect(within(pageHeader).getByText('test-bot')).toBeInTheDocument();
+  });
+
+  it('should show bot metadata', async () => {
+    withFetchSuccess();
+    renderComponent();
+    await waitForLoading();
+
+    const panel = screen.getByTestId('config-panel');
+    expect(panel).toBeInTheDocument();
+
+    expect(within(panel).getByText('test-bot')).toBeInTheDocument();
+    expect(within(panel).getByText('12h')).toBeInTheDocument();
+  });
+
+  it('should show bot roles', async () => {
+    withFetchSuccess();
+    renderComponent();
+    await waitForLoading();
+
+    const panel = screen.getByTestId('roles-panel');
```

Comment:
I wonder if those `within` queries are buying us much, given that [Testing Library discourages the use of `getByTestId`](https://testing-library.com/docs/queries/about#priority). Would it be enough it have just `expect(screen.getByText('admin')).toBeInTheDocument()`?

OTOH, if they provide function such as `within`, I can't imagine how else people are supposed to use it without doing `getByTestId`. 🤔

Reply:
This is the way (at least my one, hah!) I used the same strategy in role editor tests.